src/computability/encoding.lean

```diff
@@ -0,0 +1,160 @@
+/-
+Copyright (c) 2020 Pim Spelier, Daan van Gent. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Pim Spelier, Daan van Gent.
+-/
+
+import data.fintype.basic
+import data.num.lemmas
+import tactic
+
+/-!
+# Encodings
+
+This file contains the definition of a (finite) encoding, a map from a type to strings in an alphabet, used in defining computability by Turing machines. It also contains several examples:
+
+## Examples
+
+- `fin_encoding_nat_Γ₀₁`   : a binary encoding of ℕ in a simple alphabet.
+- `fin_encoding_nat_Γ'`    : a binary encoding of ℕ in the alphabet used for TM's.
+- `unary_fin_encoding_nat` : a unary encoding of ℕ
+- `fin_encoding_bool_Γ₀₁`  : an encoding of bool.
+-/
+
+namespace encoding
+structure encoding (α : Type) :=
+(Γ : Type)
+(encode : α → list Γ)
+(decode : list Γ → option α)
+(encodek : ∀ x, decode(encode x) = some x )
+
+structure fin_encoding (α : Type) extends encoding α :=
+(Γ_fin : fintype Γ)
+
+@[derive [inhabited,decidable_eq,fintype]]
+inductive Γ₀₁
+| bit0 | bit1
+
+@[derive [decidable_eq,fintype]]
+inductive Γ'
+| blank | bit0 | bit1 | bra | ket | comma
+
+instance inhabited_Γ' : inhabited Γ' := ⟨Γ'.blank⟩
+
+def inclusion_Γ₀₁_Γ' : Γ₀₁ → Γ'
+| Γ₀₁.bit0 := Γ'.bit0
+| Γ₀₁.bit1 := Γ'.bit1
+
+def section_Γ'_Γ₀₁ : Γ' → Γ₀₁
+| Γ'.bit0 := Γ₀₁.bit0
+| Γ'.bit1 := Γ₀₁.bit1
+| _ := arbitrary Γ₀₁
+
+lemma left_inverse_section_inclusion : function.left_inverse section_Γ'_Γ₀₁ inclusion_Γ₀₁_Γ' :=
+λ x, Γ₀₁.cases_on x rfl rfl
+
+lemma inclusion_Γ₀₁_Γ'_injective : function.injective inclusion_Γ₀₁_Γ' :=
+function.has_left_inverse.injective (Exists.intro section_Γ'_Γ₀₁ left_inverse_section_inclusion)
+
+def encode_pos_num : pos_num → list Γ₀₁
+| pos_num.one := [Γ₀₁.bit1]
+| (pos_num.bit0 n) := Γ₀₁.bit0 :: encode_pos_num n
+| (pos_num.bit1 n) := Γ₀₁.bit1 :: encode_pos_num n
+
+def encode_num : num → list Γ₀₁
+| num.zero := []
+| (num.pos n) := encode_pos_num n
+
+def encode_nat (n : ℕ) : list Γ₀₁ := encode_num n
+
+def decode_pos_num : list Γ₀₁ → pos_num
+| (Γ₀₁.bit0 :: l) := (pos_num.bit0 (decode_pos_num l))
+| (Γ₀₁.bit1 :: l) := ite (l = []) pos_num.one (pos_num.bit1 (decode_pos_num l))
+| _ := pos_num.one
+
+def decode_num : list Γ₀₁ → num := λ l, ite (l = []) num.zero $ decode_pos_num l
+
+def decode_nat : list Γ₀₁ → nat := λ l, decode_num l
+
+def encode_pos_num_nonempty (n : pos_num) : (encode_pos_num n) ≠ [] :=
```

Comment:
Declarations which are `Prop`s should be marked as `lemma` or `theorem` to keep Lean from unfolding them. Similar remarks apply to other positions in the file.